ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryRoutingFunctionalTest.java

```diff
@@ -0,0 +1,388 @@
+      .withProperty(KSQL_STREAMS_PREFIX + StreamsConfig.STATE_DIR_CONFIG, getNewStateDir())
+      .withProperty(KsqlRestConfig.LISTENERS_CONFIG, "http://localhost:8087")
+      .withProperty(KsqlRestConfig.ADVERTISED_LISTENER_CONFIG, "http://localhost:8087")
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_ENABLE_CONFIG, true)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_SEND_INTERVAL_MS_CONFIG, 600000)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_CHECK_INTERVAL_MS_CONFIG, 200)
+      .withProperty(KsqlRestConfig.KSQL_HEARTBEAT_DISCOVER_CLUSTER_MS_CONFIG, 2000)
+      .withProperty(KsqlConfig.KSQL_QUERY_PULL_ENABLE_STALE_READS, true)
+      .withProperty(KsqlConfig.KSQL_STREAMS_PREFIX + "num.standby.replicas", 1)
+      .build();
+
+  @ClassRule
+  public static final RuleChain CHAIN = RuleChain
+      .outerRule(Retry.of(3, ZooKeeperClientException.class, 3, TimeUnit.SECONDS))
+      .around(TEST_HARNESS)
+      .around(REST_APP_0)
+      .around(REST_APP_1)
+      .around(REST_APP_2);
+
+  @Rule
+  public final ExpectedException expectedException = ExpectedException.none();
+
+  @BeforeClass
+  public static void setUpClass() {
+    //Create topic with 1 partition to control who is active and standby
+    TEST_HARNESS.ensureTopics(1, USER_TOPIC);
+
+    final AtomicLong timestampSupplier = new AtomicLong(BASE_TIME);
+
+    TEST_HARNESS.produceRows(
+        USER_TOPIC,
+        USER_PROVIDER,
+        VALUE_FORMAT,
+        timestampSupplier::getAndIncrement
+    );
+
+    //Create stream
+    makeAdminRequest(
+        "CREATE STREAM " + USERS_STREAM
+            + " (" + USER_PROVIDER.ksqlSchemaString() + ")"
+            + " WITH ("
+            + "   kafka_topic='" + USER_TOPIC + "', "
+            + "   key='" + USER_PROVIDER.key() + "', "
+            + "   value_format='" + VALUE_FORMAT + "'"
+            + ");"
+    );
+  }
+
+  @Before
+  public void setUp() {
+    REST_APP_0.start();
+    REST_APP_1.start();
+    REST_APP_2.start();
+    output = KsqlIdentifierTestUtil.uniqueIdentifierName();
+
+  }
+
+  @After
+  public void cleanUp() {
+    REST_APP_0.stop();
+    REST_APP_1.stop();
+    REST_APP_2.stop();
+  }
+
+  @Test(timeout = 60000)
+  public void shouldQueryActiveWhenActiveAliveQueryIssuedToStandby() {
+    // Given:
+    final String key = Iterables.get(USER_PROVIDER.data().keySet(), 0);
+    final String sql = "SELECT * FROM " + output + " WHERE ROWKEY = '" + key + "';";
+    makeAdminRequest(
+        "CREATE TABLE " + output + " AS"
+            + " SELECT COUNT(1) AS COUNT FROM " + USERS_STREAM
+            + " GROUP BY " + USER_PROVIDER.key() + ";"
+    );
+    waitForTableRows();
+    HighAvailabilityTestUtil.waitForClusterToBeDiscovered(REST_APP_0, 3);
+    HighAvailabilityTestUtil.waitForStreamsMetadataToInitialize(
+        REST_APP_0, ImmutableList.of(host0, host1, host2), QUERY_ID);
+    ClusterFormation clusterFormation = findClusterFormation();
```

Comment:
This is common to each test.  Moving it to `setUp` would make your tests much more readable.

In general, try to make your tests read like a storey.  The test body should only have the important bits in the `Given` section, i.e. the bits that set up this test in a specific way, not the general setup.  This makes it much easier to see what it is about the test setup that is different and influencing the result.

Reply:
Done